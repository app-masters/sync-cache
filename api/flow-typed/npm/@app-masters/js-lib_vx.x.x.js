// flow-typed signature: db823a5c18f54481ff550bd09d7d2ad4
// flow-typed version: <<STUB>>/@app-masters/js-lib_v1.1.24/flow_v0.71.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@app-masters/js-lib'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@app-masters/js-lib' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@app-masters/js-lib/dist/appBootstrap' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/dateTime' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/dependencies/flexStorage' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/dependencies/rollbar' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/http' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/httpErrorHandler' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/index' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/jsError' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/notification' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/notificationCordova' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/notificationNative' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/notificationWeb' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/objHandler' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/rollbar' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/storage' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/text' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/dist/version' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/appBootstrap' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/dateTime' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/dependencies/flexStorage' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/dependencies/rollbar' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/http' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/httpErrorHandler' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/index' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/jsError' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/notification' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/notificationCordova' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/notificationNative' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/notificationWeb' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/objHandler' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/rollbar' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/storage' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/text' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/src/version' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/test/bootstrap.test' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/test/dateTime.test' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/test/http.test' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/test/localStorageMock' {
  declare module.exports: any;
}

declare module '@app-masters/js-lib/test/testConfig' {
  declare module.exports: any;
}

// Filename aliases
declare module '@app-masters/js-lib/dist/appBootstrap.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/appBootstrap'>;
}
declare module '@app-masters/js-lib/dist/dateTime.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/dateTime'>;
}
declare module '@app-masters/js-lib/dist/dependencies/flexStorage.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/dependencies/flexStorage'>;
}
declare module '@app-masters/js-lib/dist/dependencies/rollbar.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/dependencies/rollbar'>;
}
declare module '@app-masters/js-lib/dist/http.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/http'>;
}
declare module '@app-masters/js-lib/dist/httpErrorHandler.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/httpErrorHandler'>;
}
declare module '@app-masters/js-lib/dist/index.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/index'>;
}
declare module '@app-masters/js-lib/dist/jsError.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/jsError'>;
}
declare module '@app-masters/js-lib/dist/notification.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/notification'>;
}
declare module '@app-masters/js-lib/dist/notificationCordova.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/notificationCordova'>;
}
declare module '@app-masters/js-lib/dist/notificationNative.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/notificationNative'>;
}
declare module '@app-masters/js-lib/dist/notificationWeb.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/notificationWeb'>;
}
declare module '@app-masters/js-lib/dist/objHandler.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/objHandler'>;
}
declare module '@app-masters/js-lib/dist/rollbar.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/rollbar'>;
}
declare module '@app-masters/js-lib/dist/storage.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/storage'>;
}
declare module '@app-masters/js-lib/dist/text.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/text'>;
}
declare module '@app-masters/js-lib/dist/version.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/dist/version'>;
}
declare module '@app-masters/js-lib/index' {
  declare module.exports: $Exports<'@app-masters/js-lib'>;
}
declare module '@app-masters/js-lib/index.js' {
  declare module.exports: $Exports<'@app-masters/js-lib'>;
}
declare module '@app-masters/js-lib/src/appBootstrap.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/appBootstrap'>;
}
declare module '@app-masters/js-lib/src/dateTime.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/dateTime'>;
}
declare module '@app-masters/js-lib/src/dependencies/flexStorage.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/dependencies/flexStorage'>;
}
declare module '@app-masters/js-lib/src/dependencies/rollbar.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/dependencies/rollbar'>;
}
declare module '@app-masters/js-lib/src/http.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/http'>;
}
declare module '@app-masters/js-lib/src/httpErrorHandler.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/httpErrorHandler'>;
}
declare module '@app-masters/js-lib/src/index.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/index'>;
}
declare module '@app-masters/js-lib/src/jsError.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/jsError'>;
}
declare module '@app-masters/js-lib/src/notification.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/notification'>;
}
declare module '@app-masters/js-lib/src/notificationCordova.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/notificationCordova'>;
}
declare module '@app-masters/js-lib/src/notificationNative.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/notificationNative'>;
}
declare module '@app-masters/js-lib/src/notificationWeb.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/notificationWeb'>;
}
declare module '@app-masters/js-lib/src/objHandler.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/objHandler'>;
}
declare module '@app-masters/js-lib/src/rollbar.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/rollbar'>;
}
declare module '@app-masters/js-lib/src/storage.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/storage'>;
}
declare module '@app-masters/js-lib/src/text.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/text'>;
}
declare module '@app-masters/js-lib/src/version.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/src/version'>;
}
declare module '@app-masters/js-lib/test/bootstrap.test.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/test/bootstrap.test'>;
}
declare module '@app-masters/js-lib/test/dateTime.test.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/test/dateTime.test'>;
}
declare module '@app-masters/js-lib/test/http.test.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/test/http.test'>;
}
declare module '@app-masters/js-lib/test/localStorageMock.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/test/localStorageMock'>;
}
declare module '@app-masters/js-lib/test/testConfig.js' {
  declare module.exports: $Exports<'@app-masters/js-lib/test/testConfig'>;
}
