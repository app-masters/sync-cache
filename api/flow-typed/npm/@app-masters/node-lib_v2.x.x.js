declare class AMNL$ModelSequelize {
    constructor (): this;
}
declare class AMLN$Instance {
    constructor (obj: ?Object, repo: Object): this;
    spreadProperties (obj: Object): void;
    save (): Promise<void>;
    delete (): Promise<void>;
    getPropertiesById (): Promise<void>;
}

declare class AMNL$Repository<Instance> {
    static setup (sequelize: Object, modelName: string, schema: ?Object, itemInstance: Instance, modelOptions: Object): void;
    static getModel (): Object;
    static getInstance (): Instance;
    static create (obj: Object): Promise<boolean>;
    static save (obj: Object): Promise<Instance>;
    static findById (where: number | String, include: ?(string | Array<Object>), order: ?string, limit: ?number, attr: ?(Array<any> | Object)): Promise<Instance>;
    static findByIdCache (where: Object, include: ?(string | Array<Object>), order: ?string, limit: ?number, attr: ?(Array<any> | Object)): Promise<Instance>;
    static find (query: Object, include: ?(string | Array<Object>), order: ?string, limit: ?number, attr: ?(Array<any> | Object)): Promise<Array<Instance>>;
    static findCache (query: Object, include: ?(string | Array<Object>), order: ?string, limit: ?number, attr: ?(Array<any> | Object)): Promise<Array<Instance>>;
    static findOne (query: Object, include: ?(string | Array<Object>), order: ?string, limit: ?number, attr: ?(Array<any> | Object)): Promise<Instance>;
    static queryOne(sql: string): Promise<any>;
    static query(sql: string): Promise<Array<any>>;
    static update (where: Object, obj: Object): boolean;
    static delete (where: Object): Promise<boolean>;
    static updateOne (where: Object, obj: Object): Promise<Instance>;
}

declare module '@app-masters/node-lib' {
    declare export type ModelSequelize = AMNL$ModelSequelize;
    declare export type Repository<Instance> = AMNL$Repository<Instance>;
    declare export type Instance = AMLN$Instance;
    declare module.exports: {
    amAuth: any,
    amBootstreap: any,
    amController: any,
    amError:any,
    amInvite: any,
    amLeadManager: any,
    amMailing: any,
    amModel: any,
    amRdStation: any,
    amRouter: any,
    amSoftDelete: any,
    amSystem:any,
    checkVersion: any,
    dateTime: any,
    util: any,
    apiBootstrap: any,
    nodeRestful: any,
    stats: any,
    message: any,
    messageSequelize: any,
    notification: any,
    auth: any,
    modelSequelize: typeof AMNL$ModelSequelize,
    modelMongoose: any,
    jsonSchema: any,
    apiBootstrapS: any,
    express: any,
    finaleRestfull: any,
    repository: typeof AMNL$Repository,
    instance: typeof AMLN$Instance
    }
}
