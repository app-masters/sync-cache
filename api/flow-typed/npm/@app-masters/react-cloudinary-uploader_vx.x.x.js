// flow-typed signature: 1d47116953c13ec121f8437c859f2351
// flow-typed version: <<STUB>>/@app-masters/react-cloudinary-uploader_v0.1.2/flow_v0.71.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@app-masters/react-cloudinary-uploader'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@app-masters/react-cloudinary-uploader' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@app-masters/react-cloudinary-uploader/dist/index' {
  declare module.exports: any;
}

declare module '@app-masters/react-cloudinary-uploader/dist/ReactCloudinaryUploader' {
  declare module.exports: any;
}

declare module '@app-masters/react-cloudinary-uploader/src/index' {
  declare module.exports: any;
}

declare module '@app-masters/react-cloudinary-uploader/src/ReactCloudinaryUploader' {
  declare module.exports: any;
}

// Filename aliases
declare module '@app-masters/react-cloudinary-uploader/dist/index.js' {
  declare module.exports: $Exports<'@app-masters/react-cloudinary-uploader/dist/index'>;
}
declare module '@app-masters/react-cloudinary-uploader/dist/ReactCloudinaryUploader.js' {
  declare module.exports: $Exports<'@app-masters/react-cloudinary-uploader/dist/ReactCloudinaryUploader'>;
}
declare module '@app-masters/react-cloudinary-uploader/src/index.js' {
  declare module.exports: $Exports<'@app-masters/react-cloudinary-uploader/src/index'>;
}
declare module '@app-masters/react-cloudinary-uploader/src/ReactCloudinaryUploader.js' {
  declare module.exports: $Exports<'@app-masters/react-cloudinary-uploader/src/ReactCloudinaryUploader'>;
}
