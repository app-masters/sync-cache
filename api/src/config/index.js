if (process.env.NODE_ENV === 'test') {
    require('dotenv').config({path: '../.env'});
} else {
    require('dotenv').config();
}

let DATABASE = '';
if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'CI') {
    if (!process.env.DATABASE_URL) {
        throw new Error('You must create a .env file with your personal heroku database url on DATABASE_URL');
    }
    DATABASE = process.env.DATABASE_URL;
}

const session = {
    active: false
};

const rollbar = {
    accessToken: '-',
    logOnDev: false
};



module.exports = {
    test: {
        rollbar,
        database: {
            url: DATABASE
        },
        security: {
            checkClientOnDev: false,
            secret: '6d717d8cd4fc71c6f3e8e2cc5b695eace7f372c1d8ae5ddb1bfcaacebf80f543'
        },
        moip: {},
        server: {
            corsOptions: {
                origin: '*',
                methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
                preflightContinue: false,
                optionsSuccessStatus: 204,
                exposedHeaders: 'api-version, api-env, min-web-version, min-mobile-version, min-admin-version, user',
                allowedHeaders: 'content-type, Authorization, authorization, client, client-env, admin-version, mobile-version, web-version'
            },
            initialize: {
                base: '/api',
                updateMethod: 'PUT'
            }
        }
    },
    development: {
        rollbar,
        database: {
            url: DATABASE,
            forceSync: false
        },
        security: {
            checkClientOnDev: false,
            secret: '6d717d8cd4fc71c6f3e8e2cc5b695eace7f372c1d8ae5ddb1bfcaacebf80f543'
        },
        moip: {        },
        server: {
            corsOptions: {
                origin: '*',
                methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
                preflightContinue: false,
                optionsSuccessStatus: 204,
                exposedHeaders: 'api-version, api-env, min-web-version, min-mobile-version, min-admin-version, user',
                allowedHeaders: 'content-type, Authorization, authorization, client, client-env, admin-version, mobile-version, web-version'
            },
            initialize: {
                base: '/api',
                updateMethod: 'PUT'
            }
        },
        session
    },
    production: {
        rollbar,
        security: {
            checkClientOnDev: false,
            secret: '6d717d8cd4fc71c6f3e8e2cc5b695eace7f372c1d8ae5ddb1bfcaacebf80f543'
        },
        moip: {},
        server: {
            corsOptions: {
                origin: '*',
                methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
                preflightContinue: false,
                optionsSuccessStatus: 204,
                exposedHeaders: 'api-version, api-env, min-web-version, min-mobile-version, min-admin-version, user',
                allowedHeaders: 'content-type, Authorization, authorization, client, client-env, admin-version, mobile-version, web-version'
            },
            initialize: {
                base: '/api',
                updateMethod: 'PUT'
            }
        },
        session
    },
    CI: {
        rollbar,
        database: {
            url: '-'
        },
        security: {
            checkClientOnDev: false,
            secret: '6d717d8cd4fc71c6f3e8e2cc5b695eace7f372c1d8ae5ddb1bfcaacebf80f543'
        },
        server: {
            corsOptions: {
                origin: '*',
                methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
                preflightContinue: false,
                optionsSuccessStatus: 204,
                exposedHeaders: 'api-version, api-env, min-web-version, min-mobile-version, min-admin-version, user',
                allowedHeaders: 'content-type, Authorization, authorization, client, client-env, admin-version, mobile-version, web-version'
            },
            initialize: {
                base: '/api',
                updateMethod: 'PUT'
            }
        },
        moip: {},
        session
    }
};
