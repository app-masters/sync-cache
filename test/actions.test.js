import Sync from '../src/synchronization';
import { Http } from '@app-masters/js-lib';
import fetch from 'node-fetch';
import Chance from 'chance';
import { store, listener } from './redux/store';
import ACTIONS from './redux/actions/types';
import Cache from '../src/cache';
import Storage from '../src/realmStorage';
import { userReduxConfig, userCacheSchema } from './config';

const chance = new Chance();

let UserActions = null;
jest.setTimeout(20000);

let testUser = {};

describe('Sync test', () => {
    it('Storage setup', async () => {
        await Storage.setup([userCacheSchema]);
        expect(Storage.getModel()).not.toBe(null);
    });

    it('Cache setup', async () => {
        Cache.setStorage(Storage);
        expect(Cache.storage).not.toBe(null);
    });

    it('Redux setup', async () => {
        expect(store).not.toBe(null);
        expect(store.getState()).toMatchSnapshot();
    });

    it('[ONLINE] Sync setup', async () => {
        UserActions = new Sync(userReduxConfig);
        expect(UserActions).toMatchSnapshot();

        const baseUrl = 'http://localhost:3000/api';
        Http.setup('1.0.0', 'web', 'test', 'application/json');
        Http.setBaseURL(baseUrl);
        Http.setFetch(fetch);
        expect(Http.baseURL).toBe(baseUrl);
    });
    it('[ONLINE] Get users', async (finish) => {
        expect.assertions(2);
        UserActions.getObjects()(store.dispatch);
        let firstTime = true;
        listener.on(ACTIONS.USER_GET_OBJECTS, ({payload}) => {
            if (firstTime) {
                firstTime = false;
                expect(payload).toBeDefined();
            } else {
                expect(payload).toBeDefined();
                listener.removeAllListeners(ACTIONS.USER_CREATE_OBJECT);
                finish();
            }
        });
    });

    it('[ONLINE] Create user', async (finish) => {
        expect.assertions(12);
        UserActions.createObject({
            name: chance.first(),
            role: 'admin',
            localEmail: chance.email(),
            localPassword: 'validPassword'
        })(store.dispatch);
        let firstTime = true;
        listener.on(ACTIONS.USER_CREATE_OBJECT, ({payload}) => {
            if (firstTime) {
                firstTime = false;
                expect(payload).toBeDefined();
                expect(payload._id).toBeNull();
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeDefined();
                expect(payload._cacheUpdatedAt).toBeNull();
                expect(payload._cacheDeletedAt).toBeNull();
            } else {
                expect(payload).toBeDefined();
                expect(payload._id).toBeDefined();
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeNull();
                expect(payload._cacheUpdatedAt).toBeNull();
                expect(payload._cacheDeletedAt).toBeNull();
                testUser = payload;
                listener.removeAllListeners(ACTIONS.USER_CREATE_OBJECT);
                finish();
            }
        });
    });

    it('[ONLINE] Get user', async (finish) => {
        expect.assertions(12);
        UserActions.getObject(testUser._cacheId)(store.dispatch);
        let firstTime = true;
        listener.on(ACTIONS.USER_GET_OBJECT, ({payload}) => {
            if (firstTime) {
                firstTime = false;
                expect(payload).toBeDefined();
                expect(payload._id).toBe(testUser._id);
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeNull();
                expect(payload._cacheUpdatedAt).toBeNull();
                expect(payload._cacheDeletedAt).toBeNull();
            } else {
                expect(payload).toBeDefined();
                expect(payload._id).toBe(testUser._id);
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeNull();
                expect(payload._cacheUpdatedAt).toBeNull();
                expect(payload._cacheDeletedAt).toBeNull();
                listener.removeAllListeners(ACTIONS.USER_GET_OBJECT);
                finish();
            }
        });
    });

    it('[ONLINE] Update user', async (finish) => {
        expect.assertions(14);
        const role = 'user';
        const fullName = `${chance.prefix()} ${chance.first()} ${chance.last()}`;
        testUser.role = role;
        testUser.name = fullName;
        UserActions.updateObject(testUser)(store.dispatch);
        let firstTime = true;
        listener.on(ACTIONS.USER_UPDATE_OBJECT, ({payload}) => {
            if (firstTime) {
                firstTime = false;
                expect(payload).toBeDefined();
                expect(payload.role).toBe(role);
                expect(payload.name).toBe(fullName);
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeNull();
                expect(payload._cacheUpdatedAt).toBeDefined();
                expect(payload._cacheDeletedAt).toBeNull();
            } else {
                expect(payload).toBeDefined();
                expect(payload.role).toBe(role);
                expect(payload.name).toBe(fullName);
                expect(payload._cacheId).toBeDefined();
                expect(payload._cacheCreatedAt).toBeNull();
                expect(payload._cacheUpdatedAt).toBeNull();
                expect(payload._cacheDeletedAt).toBeNull();
                listener.removeAllListeners(ACTIONS.USER_UPDATE_OBJECT);
                finish();
            }
        });
    });

    it('[ONLINE] Delete user', async (finish) => {
        expect.assertions(2);
        UserActions.deleteObject(testUser)(store.dispatch);
        let firstTime = true;
        listener.on(ACTIONS.USER_DELETE_OBJECT, ({payload}) => {
            if (firstTime) {
                firstTime = false;
                expect(payload).toBeDefined();
            } else {
                expect(payload).toBeDefined();
                listener.removeAllListeners(ACTIONS.USER_DELETE_OBJECT);
                finish();
            }
        });
    });

    // OFFLINE
    it('[OFFLINE] Sync setup', async () => {
        const baseUrl = 'http://localhost:3001/api';
        Http.setBaseURL(baseUrl);
        expect(Http.baseURL).toBe(baseUrl);
    });

    it('[OFFLINE] Get users', async (finish) => {
        expect.assertions(2);
        UserActions.getObjects()(store.dispatch);
        listener.on(ACTIONS.USER_GET_OBJECTS, ({payload}) => {
            expect(payload).toBeDefined();
            listener.removeAllListeners(ACTIONS.USER_GET_OBJECTS);
            finish();
        });
    });

    it('[OFFLINE] Create user', async (finish) => {
        expect.assertions(6);
        UserActions.createObject({
            name: chance.first(),
            role: 'admin',
            localEmail: chance.email(),
            localPassword: 'validPassword'
        })(store.dispatch);
        listener.on(ACTIONS.USER_CREATE_OBJECT, ({payload}) => {
            expect(payload).toBeDefined();
            expect(payload._id).toBeNull();
            expect(payload._cacheId).toBeDefined();
            expect(payload._cacheCreatedAt).toBeDefined();
            expect(payload._cacheUpdatedAt).toBeNull();
            expect(payload._cacheDeletedAt).toBeNull();
            testUser = payload;
            listener.removeAllListeners(ACTIONS.USER_CREATE_OBJECT);
            finish();
        });
    });

    it('[OFFLINE] Get user', async (finish) => {
        expect.assertions(6);
        UserActions.getObject(testUser._cacheId)(store.dispatch);
        listener.on(ACTIONS.USER_GET_OBJECT, ({payload}) => {
            expect(payload).toBeDefined();
            expect(payload._id).toBeNull();
            expect(payload._cacheId).toBe(testUser._cacheId);
            expect(payload._cacheCreatedAt).toBeDefined();
            expect(payload._cacheUpdatedAt).toBeNull();
            expect(payload._cacheDeletedAt).toBeNull();
            listener.removeAllListeners(ACTIONS.USER_GET_OBJECT);
            finish();
        });
    });

    it('[OFFLINE] Update user', async (finish) => {
        expect.assertions(7);
        const role = 'user';
        const fullName = `${chance.prefix()} ${chance.first()} ${chance.last()}`;
        testUser.role = role;
        testUser.name = fullName;
        UserActions.updateObject(testUser)(store.dispatch);
        listener.on(ACTIONS.USER_UPDATE_OBJECT, ({payload}) => {
            expect(payload).toBeDefined();
            expect(payload.role).toBe(role);
            expect(payload.name).toBe(fullName);
            expect(payload._cacheId).toBeDefined();
            expect(payload._cacheCreatedAt).toBeDefined();
            expect(payload._cacheUpdatedAt).toBeNull();
            expect(payload._cacheDeletedAt).toBeNull();
            listener.removeAllListeners(ACTIONS.USER_UPDATE_OBJECT);
            finish();
        });
    });

    it('[OFFLINE] Delete user', async (finish) => {
        expect.assertions(1);
        console.log(testUser);
        UserActions.deleteObject(testUser)(store.dispatch);
        listener.on(ACTIONS.USER_DELETE_OBJECT, ({payload}) => {
            expect(payload).toBeDefined();
            listener.removeAllListeners(ACTIONS.USER_DELETE_OBJECT);
            finish();
        });
    });
    const testNumber = 10;
    it(`[OFFLINE] Create ${testNumber} users`, async (finish) => {
        expect.assertions(testNumber);
        let successCount = 0;
        listener.on(ACTIONS.USER_CREATE_OBJECT, ({payload}) => {
            expect(payload).toBeDefined();
            successCount++;
            if (successCount === testNumber) {
                listener.removeAllListeners(ACTIONS.USER_CREATE_OBJECT);
                finish();
            }
        });
        for (let index of Array.from(Array(testNumber), (_, i) => i + 1)) {
            UserActions.createObject({
                name: `${chance.prefix()} ${chance.first()} ${chance.last()}`,
                role: 'admin',
                localEmail: chance.email(),
                localPassword: 'validPassword'
            })(store.dispatch);
        }
    });

    it(`[OFFLINE] ${testNumber} users to sync`, async () => {
        const usersToCreate = await UserActions.getObjectsToCreate();
        expect(usersToCreate).toBeDefined();
        expect(usersToCreate).toHaveLength(testNumber);
    });

    it(`[ONLINE] Sync ${testNumber} users`, async () => {
        const baseUrl = 'http://localhost:3000/api';
        Http.setBaseURL(baseUrl);
        const createdUsers = await UserActions.syncCreatedObjects();
        expect(createdUsers).toBeDefined();
        expect(createdUsers).toHaveLength(testNumber);
    });

});