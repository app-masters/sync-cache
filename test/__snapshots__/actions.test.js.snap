// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync test [ONLINE] Sync setup 1`] = `
Synchronization {
  "config": Object {
    "cacheStrategy": "CacheOnline",
    "conflictRule": [Function],
    "createSuffix": "",
    "deleteSuffix": "",
    "endPoint": "/users/",
    "prepareToClient": [Function],
    "prepareToServer": [Function],
    "primaryKey": "_id",
    "typePrefix": "USER",
    "updateSuffix": "",
    "validateObject": [Function],
  },
}
`;

exports[`Sync test Redux setup 1`] = `
Object {
  "error": false,
  "input": Object {},
  "item": null,
  "items": Array [],
  "loading": false,
  "loadingCache": false,
  "loadingOnline": false,
}
`;
